DataFormats/Provenance/test/HardwareResourcesDescription_t.cpp

```diff
@@ -0,0 +1,76 @@
+
+#include "DataFormats/Provenance/interface/HardwareResourcesDescription.h"
+#include "FWCore/Utilities/interface/EDMException.h"
+
+TEST_CASE("HardwareResourcesDescription", "[HardwareResourcesDescription]") {
+  SECTION("Construction from empty string") {
+    CHECK(edm::HardwareResourcesDescription("") == edm::HardwareResourcesDescription());
+  }
+
+  SECTION("Default construction") {
+    edm::HardwareResourcesDescription resources;
+    CHECK(edm::HardwareResourcesDescription(resources.serialize()) == resources);
+  }
+
+  SECTION("Microarchitecture") {
+    edm::HardwareResourcesDescription resources;
+    resources.microarchitecture = "x86-64-v3";
+    CHECK(edm::HardwareResourcesDescription(resources.serialize()) == resources);
+  }
+
+  SECTION("CPU models") {
+    edm::HardwareResourcesDescription resources;
+    resources.cpuModels = {"Intel something", "AMD something else"};
+    CHECK(edm::HardwareResourcesDescription(resources.serialize()) == resources);
+  }
+
+  SECTION("accelerators") {
+    edm::HardwareResourcesDescription resources;
+    resources.selectedAccelerators = {"cpu", "gpu"};
+    CHECK(edm::HardwareResourcesDescription(resources.serialize()) == resources);
+  }
+
+  SECTION("GPU models") {
+    edm::HardwareResourcesDescription resources;
+    resources.gpuModels = {"NVIDIA something", "NVIDIA something else"};
+    CHECK(edm::HardwareResourcesDescription(resources.serialize()) == resources);
+  }
+
+  SECTION("All fields") {
+    edm::HardwareResourcesDescription resources;
+    resources.microarchitecture = "x86-64-v3";
+    resources.cpuModels = {"Intel something", "AMD something else"};
+    resources.selectedAccelerators = {"cpu", "gpu"};
+    resources.gpuModels = {"NVIDIA something", "NVIDIA something else"};
+    CHECK(edm::HardwareResourcesDescription(resources.serialize()) == resources);
+  }
+
+  SECTION("Serialization has additional things (forward compatibility)") {
+    edm::HardwareResourcesDescription resources, resources2;
+    resources.microarchitecture = "x86-64-v3";
+    resources.cpuModels = {"Intel something", "AMD something else"};
+    resources.selectedAccelerators = {"cpu", "gpu"};
+    resources.gpuModels = {"NVIDIA something", "NVIDIA something else"};
+    resources2.microarchitecture = "this";
+    resources2.cpuModels = {"is"};
+    resources2.selectedAccelerators = {"something"};
+    resources2.gpuModels = {"else"};
+    auto const serial = resources.serialize() + resources2.serialize();
+    CHECK(edm::HardwareResourcesDescription(serial) == resources);
+  }
+
+  SECTION("Error cases") {
+    SECTION("Invalid serialized string") {
+      CHECK_THROWS_AS(edm::HardwareResourcesDescription("foo"), edm::Exception);
+
+      edm::HardwareResourcesDescription resources;
+      resources.microarchitecture = "x86-64-v3";
+      auto serialized = resources.serialize();
+      serialized.back() = ',';
+      CHECK_THROWS_AS(edm::HardwareResourcesDescription(serialized), edm::Exception);
```

Comment:
Just to clarify, so in this case, the serialized string ends with the wrong character, causing the failure. The case that follows this fails because it is missing the closing character.

Reply:
More or less. Another way to phrase would be
* this case fails because the last container (that contains `,`) is missing the closing character
* the following case fails because the container is missing